pkg/apis/resource/validation/validation.go

```diff
@@ -788,21 +926,121 @@ func validateDevice(device resource.Device, fldPath *field.Path, sharedCounterTo
+		func(deviceMixinRef string, fldPath *field.Path) field.ErrorList {
+			return validateDeviceMixinRef(deviceMixinRef, fldPath, deviceMixinNames)
+		}, fldPath.Child("includes"))...)
+
+	// Check the total number of attributes and capacities in flattened device.
+	flattenedAttributes, flattenedCapacities := flattenAttributesAndCapacities(device, deviceMixins)
+	numFlattenedAttributesAndCapacities := len(flattenedAttributes) + len(flattenedCapacities)
+	if numFlattenedAttributesAndCapacities > resource.ResourceSliceMaxAttributesAndCapacitiesPerDeviceAfterMixins {
+		allErrs = append(allErrs, field.Invalid(fldPath, numFlattenedAttributesAndCapacities, fmt.Sprintf("the total number of attributes and capacities in a device after mixins have been applied must not exceed %d", resource.ResourceSliceMaxAttributesAndCapacitiesPerDeviceAfterMixins)))
+	}
+
 	return allErrs
 }
 
-func validateDeviceCounterConsumption(deviceCounterConsumption resource.DeviceCounterConsumption, fldPath *field.Path) field.ErrorList {
+func flattenAttributesAndCapacities(device resource.Device, deviceMixins []resource.DeviceMixin) (map[resource.QualifiedName]resource.DeviceAttribute, map[resource.QualifiedName]resource.DeviceCapacity) {
+	if len(device.Includes) == 0 || len(deviceMixins) == 0 {
+		return device.Attributes, device.Capacity
+	}
+
+	attributes := make(map[resource.QualifiedName]resource.DeviceAttribute)
+	capacities := make(map[resource.QualifiedName]resource.DeviceCapacity)
+	for qname, attr := range device.Attributes {
+		attributes[qname] = attr
+	}
+	for qname, cap := range device.Capacity {
+		capacities[qname] = cap
+	}
+	for _, mixinRef := range device.Includes {
+		i := slices.IndexFunc(deviceMixins, func(dm resource.DeviceMixin) bool {
+			return dm.Name == mixinRef
+		})
+		if i < 0 {
+			continue
+		}
+		mixin := deviceMixins[i]
+		for qname, attr := range mixin.Attributes {
+			attributes[qname] = attr
+		}
+		for qname, cap := range mixin.Capacity {
+			capacities[qname] = cap
+		}
+	}
+	return attributes, capacities
+}
+
+func validateDeviceMixinRef(deviceMixinRef string, fldPath *field.Path, deviceMixinNames sets.Set[string]) field.ErrorList {
 	var allErrs field.ErrorList
+	if !deviceMixinNames.Has(deviceMixinRef) {
+		allErrs = append(allErrs, field.Invalid(fldPath, deviceMixinRef, "must reference a device mixin defined in the ResourceSlice"))
+	}
+	return allErrs
+}
 
+func validateDeviceCounterConsumption(deviceCounterConsumption resource.DeviceCounterConsumption, fldPath *field.Path, deviceCounterConsumptionMixinToCounterMap,
+	sharedCounterToCounterNames map[string]sets.Set[string]) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Make sure a valid CounterSet is referenced.
 	if len(deviceCounterConsumption.CounterSet) == 0 {
 		allErrs = append(allErrs, field.Required(fldPath.Child("counterSet"), ""))
 	}
-	if len(deviceCounterConsumption.Counters) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("counters"), ""))
```

Comment:
is this field still alpha-level in 1.34? if so, this relaxed validation is fine. if not, we have to make sure we don't break 1.33 servers with relaxed data in this field